[quant] add reduce_range option to FusedMovingAvgFakeQuantize module

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62702
* __->__ #62863

Summary:
To make this consistent with other observers, add reduce_range option that can be used to update quant_min/max.

Updated the observer code to move the calculation of qparams to the init function so that it is done once at observer initialization and not during calculate_qparams that gets called during convert.

Test Plan:
python test/test_quantization.py test_fused_mod_reduce_range

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30146602](https://our.internmc.facebook.com/intern/diff/D30146602)

supriyar committed:
[quant] add reduce_range option to FusedMovingAvgFakeQuantize module
Summary:
To make this consistent with other observers, add reduce_range option that can be used to update quant_min/max

Test Plan:
python test/test_quantization.py test_fused_mod_reduce_range

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30146602).

raghuramank100 commented on torch/quantization/utils.py:
```diff
@@ -133,3 +134,49 @@ def check_min_max_valid(min_val: torch.Tensor, max_val: torch.Tensor) -> bool:
     return True
+
+
+def calculate_qmin_qmax(quant_min: int, quant_max: int, has_customized_qrange: bool, dtype: torch.dtype,
```
Why is the _ removed? Do we see this as a user facing function?

supriyar replied:
Kept it consistent with the rest of functions in this file. Since it is in utils it shouldn't be a user facing function.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -193,8 +193,8 @@ def __init__(
             self._validate_qmin_qmax(quant_min, quant_max)
-        self.quant_min = quant_min
-        self.quant_max = quant_max
+        self.quant_min, self.quant_max = \
```
Good to see the calculation of quant min and quant max being done once now

raghuramank100 commented on test/quantization/core/test_workflow_module.py:
```diff
@@ -1096,6 +1096,12 @@ def test_fused_mod_per_channel(self):
             torch.testing.assert_allclose(mod.state_dict()['activation_post_process.max_val'], running_max_op)
 
+    def test_fused_mod_reduce_range(self):
+        obs = FusedMovingAvgObsFakeQuantize(quant_min=0, quant_max=255, dtype=torch.quint8, reduce_range=True)
```
Should the default version have reduce_range to be True, as we default to fbgemm?. Currently it looks like it does not that.

vkuzo commented on torch/quantization/fake_quantize.py:
```diff
@@ -280,6 +280,8 @@ def __init__(
         self.register_buffer("observer_enabled", torch.tensor([0], dtype=torch.long))
         self.is_symmetric_quant = _is_symmetric_quant(self.activation_post_process.qscheme)
 
+        self.quant_min, self.quant_max = self.activation_post_process.quant_min, self.activation_post_process.quant_max
```
is this guaranteed to be executed after `self.activation_post_process` is initialized, in all cases?

vkuzo replied:
qq, does it make sense to take `quant_min/quant_max` as constructor arguments, if they are overriden by this line of code?

supriyar replied:
> is this guaranteed to be executed after `self.activation_post_process` is initialized, in all cases?

Yes, it should be since we call the constructer of the observer before this and I moved the quant_min/max calculation to the observer init function

> does it make sense to take quant_min/quant_max as constructor arguments, if they are overriden by this line of code?

It doesn't. I noticed this is how the current FakeQuantize modules do it too, even though we have a quant_min/max argument passed in the ultimate thing that gets used is the one set by the observer. It feels quite confusing to me.

supriyar committed:
Update on "[quant] add reduce_range option to FusedMovingAvgFakeQuantize module"


Summary:
To make this consistent with other observers, add reduce_range option that can be used to update quant_min/max.

Updated the observer code to move the calculation of qparams to the init function so that it is done once at observer initialization and not during calculate_qparams that gets called during convert.

Test Plan:
python test/test_quantization.py test_fused_mod_reduce_range

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30146602](https://our.internmc.facebook.com/intern/diff/D30146602)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30146602).

supriyar committed:
Update on "[quant] add reduce_range option to FusedMovingAvgFakeQuantize module"


Summary:
To make this consistent with other observers, add reduce_range option that can be used to update quant_min/max.

Updated the observer code to move the calculation of qparams to the init function so that it is done once at observer initialization and not during calculate_qparams that gets called during convert.

Test Plan:
python test/test_quantization.py test_fused_mod_reduce_range

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30146602](https://our.internmc.facebook.com/intern/diff/D30146602)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30146602).
